File: ex/clean/src/main/kotlin/com/ktb/shop/adapters/notification/LogNotifier.kt
package com.ktb.shop.adapters.notification

import com.ktb.shop.entities.Customer
import com.ktb.shop.usecases.port.Notifier
import jakarta.inject.Singleton
import org.slf4j.LoggerFactory

// GATEWAY for the Notifier port. Swap for a real SMTP/SendGrid adapter later —
// the interactor (RegisterCustomerInteractor) does not change.
@Singleton
class LogNotifier : Notifier {
    private val log = LoggerFactory.getLogger(javaClass)

    override fun welcome(customer: Customer) {
        log.info("welcome email -> {} ({})", customer.email.value, customer.name)
    }
}

File: ex/clean-go/internal/adapters/notification/log_notifier.go
package notification

import (
	"context"
	"log/slog"

	"github.com/ktb/clean-shop/internal/entities"
)

// GATEWAY implementation for the Notifier output port. Swap for a real
// SMTP/SendGrid adapter later — the interactor does not change.
type LogNotifier struct{}

func NewLogNotifier() *LogNotifier { return &LogNotifier{} }

func (n *LogNotifier) Welcome(_ context.Context, customer entities.Customer) {
	slog.Info("welcome email", "to", customer.Email.String(), "name", customer.Name)
}
